cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -0,0 +1,354 @@
+               && isReplicaCountAboveBalanceLowerLimitAfterChange(sourceBroker, sourceBroker.leaderReplicas().size(), REMOVE))
+               ? ACCEPT : REPLICA_REJECT;
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  @Override
+  public ClusterModelStatsComparator clusterModelStatsComparator() {
+    return new LeaderReplicaDistributionGoalStatsComparator();
+  }
+
+  @Override
+  public String name() {
+    return LeaderReplicaDistributionGoal.class.getSimpleName();
+  }
+
+  /**
+   * Rebalance the given broker without violating the constraints of the current goal and optimized goals.
+   *
+   * @param broker         Broker to be balanced.
+   * @param clusterModel   The state of the cluster.
+   * @param optimizedGoals Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization -- e.g. excluded topics.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker,
+                                    ClusterModel clusterModel,
+                                    Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) {
+    LOG.debug("Rebalancing broker {} [limits] lower: {} upper: {}.", broker.id(), _balanceLowerLimit, _balanceUpperLimit);
+    int numLeaderReplicas = broker.leaderReplicas().size();
+    boolean requireLessLeaderReplicas = broker.isAlive() ? numLeaderReplicas > _balanceUpperLimit : numLeaderReplicas > 0;
+    boolean requireMoreLeaderReplicas = broker.isAlive() && numLeaderReplicas < _balanceLowerLimit;
+    // Update broker ids over the balance limit for logging purposes.
+    if (requireLessLeaderReplicas
+        && rebalanceByMovingLeadershipOut(broker, clusterModel, optimizedGoals, optimizationOptions)
+        && rebalanceByMovingLeaderReplicasOut(broker, clusterModel, optimizedGoals, optimizationOptions)) {
+      _brokerIdsAboveBalanceUpperLimit.add(broker.id());
+      LOG.debug("Failed to sufficiently decrease leader replica count in broker {}. Leader replicas: {}.",
+                broker.id(), broker.leaderReplicas().size());
+    } else if (requireMoreLeaderReplicas
+               && rebalanceByMovingLeadershipIn(broker, clusterModel, optimizedGoals, optimizationOptions)
+               && rebalanceByMovingLeaderReplicasIn(broker, clusterModel, optimizedGoals, optimizationOptions)) {
+      _brokerIdsUnderBalanceLowerLimit.add(broker.id());
+      LOG.debug("Failed to sufficiently increase leader replica count in broker {}. Leader replicas: {}.",
+                broker.id(), broker.leaderReplicas().size());
+    }
+    // If the broker is dead and currently self healing dead brokers only, then generate proposals to move all the
+    // remaining replicas off this broker.
+    if (_selfHealingDeadBrokersOnly && !broker.isAlive() && !broker.replicas().isEmpty()) {
```

Comment:
This flow seems to be contradicting the other goals -- why don't we move this logic to `rebalanceByMovingLeadershipOut`?

Reply:
Looks this cause a lot of confusion... 
My origin idea is to make `rebalanceByMovingLeadershipOut` easy-to-induction, I will merge the logic into `rebalanceByMovingLeadershipOut`.